Data/HashMap/Base.hs

```diff
@@ -1932,12 +1932,12 @@ updateWith# f k0 ary0 = go k0 ary0 0 (A.length ary0)
                      | otherwise -> go k ary (i+1) n
 {-# INLINABLE updateWith# #-}
 
-updateOrSnocWith :: Eq k => (v -> v -> v) -> k -> v -> A.Array (Leaf k v)
+updateOrSnocWith :: Eq k => (v -> v -> (# v #)) -> k -> v -> A.Array (Leaf k v)
                  -> A.Array (Leaf k v)
 updateOrSnocWith f = updateOrSnocWithKey (const f)
 {-# INLINABLE updateOrSnocWith #-}
 
-updateOrSnocWithKey :: Eq k => (k -> v -> v -> v) -> k -> v -> A.Array (Leaf k v)
+updateOrSnocWithKey :: Eq k => (k -> v -> v -> (# v #)) -> k -> v -> A.Array (Leaf k v)
```

Comment:
Maybe add a reference to David's explanation of the unboxed-tuple trick here?

Reply:
We use the unboxed tuple trick in a bunch of places in HashMap, so it seems "expected knowledge" for working with this package.

Reply:
Perhaps we should raise a general doc issue for this trick then. It's uncommon knowledge I think 😄

Reply:
I think a blog post explaining the trick would be great. I wouldn't document it in the context of unordered-containers, but try and make it a general technique, and just link to it from this project.

Reply:
Yeah, a blog post would be great. I have opened https://github.com/haskell-unordered-containers/unordered-containers/issues/262 as a reminder.

Reply:
I would call this `updateOrSnocWithKey#`, and define `updateOrSnocWithKey` to use it in the boring way. As it is, this function can only be used for the *lazy* operations. Should we extend it to let it work for strict ones too? The most semantically pleasing version would look like this:

```haskell
updateOrSnocWith#
  :: Eq k
  => ((# (# #) | v #) -> (# v #))
  -> k
  -> A.Array (Leaf k v)
  -> A.Array (Leaf k v)
  -> A.Array (Leaf k v)
```

This takes a function that we pass either `(# (# #) | #)` (meaning the key was not present) or `(# | v #)` (the value tied to the key in the map) and produces an unboxed unary tuple for the new value.

Unfortunately, we support GHC versions that don't have unboxed sums. We can work around that in one of two ways:

```haskell
updateOrSnocWith#
  :: Eq k
  => (Maybe v -> (# v #))
  -> k
  -> A.Array (Leaf k v)
  -> A.Array (Leaf k v)
  -> A.Array (Leaf k v)
```

```haskell
updateOrSnocWith#
  :: Eq k
  => (v -> (# v #))
  -> k
  -> ((# #) -> (# v #))
  -> A.Array (Leaf k v)
  -> A.Array (Leaf k v)
  -> A.Array (Leaf k v)
```

Another issue is evaluating the whole `v -> v` vs. `v -> v -> v` sort of approach, which has to do with what closures we might have to create in different contexts. This will come down to poring over Core and doing some benchmarking. This is actually true whether we try to unify strict and lazy or not. Personally, I'd *love* to be able to unify as much as we can, if we can get enough of the right things to inline the way we want to avoid hurting performance.

Final note: if we're careful enough about inlining, we might well be able to just use boxed everything and still not pay for it. That means making sure that the underlying implementation inlines into the `Strict` or `Lazy` version as soon as the wrapped-up function is passed in, while that `Strict` or `Lazy` version itself remains (at least) `INLINABLE`.

Reply:
Is it useful to add both `updateOrSnocWith#` and `updateOrSnocWith`? For user code, where we expose it, I agree that having a "friendly" wrapper is valuable. But given we have `updateOrSnocWith` and `updateOrSnocWithKey` that seems like a lot of intermediate helper functions that don't get used a whole lot.

My inclination would be to keep this patch (which fixes a space leak) a small self-contained space leak fix, and then follow up with the larger unboxed sums and looking at the Core after. It feels like a subsequent refactor, and one that I'm not well placed to make.